x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineFilters.java

```diff
@@ -84,7 +85,10 @@ protected LogicalPlan rule(Filter filter) {
         } else if (child instanceof OrderBy orderBy) {
             // swap the filter with its child
             plan = orderBy.replaceChild(filter.with(orderBy.child(), condition));
-        } else if (child instanceof Join join) {
+        } else if (child instanceof Join join && child instanceof InlineJoin == false) {
+            // TODO: could we do better here about pushing down filters for inlinestats?
+            // See also https://github.com/elastic/elasticsearch/issues/127497
+            // Push down past INLINESTATS if the condition is on the groupings
```

Comment:
I believe we can do better, but the stub relation approach of `InlineJoin` is getting in the way.

The StubRelation approach also hinges on the subtle idea that the same name ids occuring on the left and right hand side of the join is okay, but I have doubts about that.

Since we're not physically re-using the left hand side output to compute the stats on the right, but we instead run 2 separate (but dependent) queries, I believe it may be simpler to just "copy" the left hand side in place of the stub; this would allow for independent optimization more easily and be more honest, although I haven't thought through all the implications of what I just said.

Reply:
Indeed, StubRelation needs a rethinking especially related to Mapper. Thanks @alex-spies